# Convolution Neural Networks

For this project the images from the MNIST database will be used. MNIST contains images of handwritten digits from 0 to 9 and includes 60,000 grayscale training images of size 28*28 pixels and 10,000 test images of the same dimensions and features(K=10 classes). It also contains labels for all images.
Solving the MNIST database problem involves partitioning the data into classes in a way that all the contents of each cluster belong to the same class. More specifically, each cluster and, by extension, class, will consist of images containing a handwritten digit from 0 to 9. A Convolution Neural Network(CNN) will be train so that it can separate the data and, given an input-image, can "recognize ” the class in which it belongs(that is, to identify the number included in the image).

In [ ]:
# Installation of the libraries keras και tensorflow
!pip install keras
!pip install tensorflow

In [1]:
# Import of the necessary libraries and the MNIST dataset
from keras import layers
from keras import models
from keras.datasets import mnist
from keras.utils import to_categorical

In [2]:
# Loading the dataset from the keras library datasets by importing mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

### Preprocessing of the data

Defining of the dimensions of each image

In [5]:
# The input is 256 28*28 pixels gray-scale pictures
# 60000 instances for training and 10000 for testing
train_images = train_images.reshape((60000, 28, 28, 1))
test_images = test_images.reshape((10000, 28, 28, 1))

It is more common to use 32-bit precision when training a neural network, so at this point the training data should be converted to 32-bit floats by dividing by 255. As for dividing by 255, this is the maximum value of a byte, so this will ensure that the input features scale between 0.0 and 1.0. This is not mandatory, but it is usually preferable to have input attributes on this scale.

In [6]:
# Setting the type of train_images and test_images to float32 dividing by 255
train_images = train_images.astype('float32')/255
test_images = test_images.astype('float32')/255
train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

In [7]:
# Implementation of the CNN model
model = models.Sequential()

### Layer creation

Defining of the dimensions of the filter to be used and the dimension of the pooling window for sub-sampling the data

In [9]:
# Using the Conv2D layer from the keras library with 32 neurons in the first layer and with a filter size of 3*3
model.add(layers.Conv2D(32, (3, 3), activation = 'relu', padding = 'same', input_shape = (28, 28, 1)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation = 'relu', padding = 'valid'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation = 'relu'))

# The first Conv2D layer has a padding = same, while the second has a padding = valid
# (padding = valid is equivalent to no padding
#padding = same means that the size of the output of this layer must be equal to the size of the input)

In total, the implementation of padding, convolutional layer with ReLU as the activation function and a max-pooling layer is being used

In [10]:
# Using the summary function, all layer names, parameters and shapes are displayed
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 28, 28, 32)        320       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 14, 14, 32)       0         
 )                                                               
                                                                 
 conv2d_1 (Conv2D)           (None, 12, 12, 64)        18496     
                                                                 
 max_pooling2d_1 (MaxPooling  (None, 6, 6, 64)         0         
 2D)                                                             
                                                                 
 conv2d_2 (Conv2D)           (None, 4, 4, 64)          36928     
                                                                 
Total params: 55,744
Trainable params: 55,744
Non-traina

Addition of a flattened layer along with two fully connected layers

In [11]:
# The final fully connected layer should have enough neurons equal to the number of classes
# we want to classify along with a SoftMax activation function
model.add(layers.Flatten())
model.add(layers.Dense(64, activation = 'relu'))
model.add(layers.Dense(10, activation = 'softmax'))

In [12]:
# Using of the summary function again
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 28, 28, 32)        320       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 14, 14, 32)       0         
 )                                                               
                                                                 
 conv2d_1 (Conv2D)           (None, 12, 12, 64)        18496     
                                                                 
 max_pooling2d_1 (MaxPooling  (None, 6, 6, 64)         0         
 2D)                                                             
                                                                 
 conv2d_2 (Conv2D)           (None, 4, 4, 64)          36928     
                                                                 
 flatten (Flatten)           (None, 1024)              0

### Compiling the model using the compile() function

In [13]:
# Use of the RMSprop optimizer, the Multi-Class Cross-Entropy Loss function
# and the Accuracy metric to evaluate the model
model.compile(optimizer = 'rmsprop', loss = 'categorical_crossentropy', metrics = ['accuracy'])

In [14]:
# Feeding of the model and settting of the number of epochs(5 in this case)
# Settting the sample size to 64
model.fit(train_images, train_labels, epochs = 5, batch_size = 64)

Epoch 1/5
938/938 [==============================] - 28s 28ms/step - loss: 0.1601 - accuracy: 0.9497
Epoch 2/5
938/938 [==============================] - 25s 27ms/step - loss: 0.0436 - accuracy: 0.9865
Epoch 3/5
938/938 [==============================] - 25s 27ms/step - loss: 0.0291 - accuracy: 0.9907
Epoch 4/5
938/938 [==============================] - 25s 27ms/step - loss: 0.0224 - accuracy: 0.9927
Epoch 5/5
938/938 [==============================] - 25s 27ms/step - loss: 0.0168 - accuracy: 0.9950


In [15]:
# Model check
test_loss, test_acc = model.evaluate(test_images, test_labels)

313/313 [==============================] - 3s 8ms/step - loss: 0.0277 - accuracy: 0.9915


In [16]:
# Displaying the accuracy
print('Test accuracy:', test_acc)

Test accuracy: 0.9915000200271606


Our model displayed an accuracy of 99.15%! An excellent level of accuracy, nearly similar to the levels of accuracy one would expect from a human.